[Ready] String encoding fixes

lozhnikov opened the pull request:
The PR absorbs #1814.

- [x] The API is stable.
  - [x] Benchmarks for `boost::optional` as the output of the tokenizers are done.
  - [x] The tokenizers don't return `boost::optional` (not needed).
- [x] The string encoding policies API is stable.
- [x] All algorithms work fine.
- [x] Everything looks great.

lozhnikov commented:
@jeffin143 The string encoding policies API is stable, so you can fix your implementation of the bag of words algorithm and Tf-Idf.

lozhnikov commented:
# Background

Currently each tokenizer has a special method (`IsTokenEmpty()`) that returns `true` if the given token is empty. Hence we can't use lambda expressions as tokenizers since the tokenizers return various value types. Modern C++ has an obvious solution: `std::optional`. I tried to investigate if we can use `boost::optional` as the return value type of the tokenizers (i.e. the tokenizer returns `nullopt` if the next token is empty).

Source code: [boost-optional-tests.zip](https://github.com/mlpack/mlpack/files/3449020/boost-optional-tests.zip)

I used the same compilers, the same compiler options, the same benchmark options and the same input data as [here](https://github.com/mlpack/mlpack/pull/1814#issuecomment-514687037).

**All figures show the time of the described tests**.

**The figures related to the same tests have the same scale**.

# Benchmarks
## Tokenizer comparison. Part 1.
### These tokenizers split the given string using the given set of delimiters

There are three tests for two different tokenizers.
- The first tokenizer returns `boost::string_view`.
- The second tokenizer returns `boost::optional<boost::string_view>`.

The first test just splits a number of strings into tokens. The second test does the same and saves the result into `std::vector`. The third test extracts tokens and saves them into `std::unordered_set`.

### g++ 9.1.1

![g++_tokenizers_benchmark](https://user-images.githubusercontent.com/10090844/62168417-34f74a80-b32e-11e9-944d-34ef95544eef.png)

### clang++ 8.0

![clang++_tokenizers_benchmark](https://user-images.githubusercontent.com/10090844/62168431-3b85c200-b32e-11e9-9a70-45f252134a65.png)

### clang++ 8.0 and libc++

![clang++_libc++_tokenizers_benchmark](https://user-images.githubusercontent.com/10090844/62168439-417ba300-b32e-11e9-8913-f7b28f4cb337.png)

## Tokenizer comparison. Part 2.
### The tokenizers below extract a character at a time

There are three tests for two different tokenizers.
- The first tokenizer returns `int`.
- The second tokenizer returns `boost::optional<int>`.

The first test just splits a number of strings into tokens. The second test does the same and saves the result into `std::vector`. The third test extracts tokens and saves them into `std::unordered_set`.

### g++ 9.1.1

![g++_char_extract_tokenizers_benchmark](https://user-images.githubusercontent.com/10090844/62168460-4f312880-b32e-11e9-8578-9493730a5b06.png)

### clang++ 8.0

![clang++_char_extract_tokenizers_benchmark](https://user-images.githubusercontent.com/10090844/62168469-55bfa000-b32e-11e9-98f9-4c6037b7111d.png)

### clang++ 8.0 and libc++

![clang++_libc++_char_extract_tokenizers_benchmark](https://user-images.githubusercontent.com/10090844/62168476-59ebbd80-b32e-11e9-85ba-4df45a7d40be.png)

## Encode() tests. Part 1.
### First consider the `Encode()` function that accepts a tokenizer which splits the given string using the given set of delimiters

There are two tests for two different tokenizers.
- The first tokenizer returns `boost::string_view`.
- The second tokenizer returns `boost::optional<boost::string_view>`.

The first test encodes a number of strings in two passes:
1. Pass 1: Extract all tokens and add them to a dictionary (`std::unordered_map`).
2. Pass 2: Encode all tokens using the dictionary and save them somewhere.

The second test encodes a number of strings in one pass:
1. For all tokens: extract a token, add it to a dictionary (`std::unordered_map`), encode the token and save the result.

### g++ 9.1.1

![g++_encode_benchmark](https://user-images.githubusercontent.com/10090844/62168481-5e17db00-b32e-11e9-87bf-39d6f7bf58ae.png)

### clang++ 8.0

![clang++_encode_benchmark](https://user-images.githubusercontent.com/10090844/62168490-6243f880-b32e-11e9-818f-6bfb00f9e6d3.png)

### clang++ 8.0 and libc++

![clang++_libc++_encode_benchmark](https://user-images.githubusercontent.com/10090844/62168510-68d27000-b32e-11e9-8905-c2d51bb1f93a.png)

## Encode() tests. Part 2.
### Now consider the `Encode()` function that accepts a tokenizer which extracts a character at a time

There are two tests for two different tokenizers.
- The first tokenizer returns `int`.
- The second tokenizer returns `boost::optional<int>`.

The first test encodes a number of strings in two passes:
1. Pass 1: Extract all tokens and add them to a dictionary (`std::array`).
2. Pass 2: Encode all tokens using the dictionary and save them somewhere.

The second test encodes a number of strings in one pass:
1. For all tokens: extract a token, add it to a dictionary (`std::array`), encode the token and save the result.

### g++ 9.1.1

![g++_char_encode_benchmark](https://user-images.githubusercontent.com/10090844/62168526-6f60e780-b32e-11e9-9fbc-7471f4320dbb.png)

### clang++ 8.0

![clang++_char_encode_benchmark](https://user-images.githubusercontent.com/10090844/62168543-75ef5f00-b32e-11e9-86d6-5095162b2627.png)

### clang++ 8.0 and libc++

![clang++_libc++_char_encode_benchmark](https://user-images.githubusercontent.com/10090844/62168552-7a1b7c80-b32e-11e9-984f-883fd4c60c83.png)

# Conclusion

- The very first figure (Tokenizers, part 1, g++ 9.1.1) looks like an anomaly. I think `g++` optimized something out, but I didn't find out why it happens.
- We can neglect the performance difference between tokenizers that return `boost::string_view` and `boost::optional<boost::string_view>`.
- The tokenizer that returns `boost::optional<int>` works much slower than the one that returns `int`. However we should take into account that the dictionary encoding algorithm is pretty simple in case of encoding individual characters. @jeffin143 Perhaps I will repeat the benchmark as soon as Jeffin implements other algorithms.
- I think it is reasonable to keep the present API i.e. the `Encode()` function has the `TokenizerType`  template argument  and each tokenizer has the `IsTokenEmpty()` method. This is rather robust approach, and we can switch to `boost::optional` without external changes anytime.

Thanks for the attention!

lozhnikov commented:
AppVeyor throws the following error:
`multiple versions of a defaulted special member functions are not allowed`.

Looks like it's a MSVS bug (see [stackoverflow](https://stackoverflow.com/questions/36657243/c-multiple-versions-of-a-defaulted-special-member-functions-error-in-msvc-2)). Old versions of Visual Studio (however, MSVS2017 build failed too) doesn't compile the following code:
```c++
class Foo
{
  template<typename ...Args>
  Foo(Args&& ...args);

  // We has to declare explicitly the copy-constructor since the parameter-pack
  // constructor hides it.
  Foo(const Foo&) = default;
};
```
@rcurtin, @zoq Is there an easy fix (except adding the implementation of the required constructors)?

The constructors are pretty simple, but `= default` looks better than the explicit implementation.

rcurtin commented:
@lozhnikov I took a look... not too surprising that MSVS fails something that it shouldn't even after they said they fixed it.  We've definitely had a lot of those issues before.  I agree that it's uglier but I think your solution in 2a5464e is probably unfortunately the right way to go...

`std::optional` is C++17, which might be a little hard.  I don't have too much of a problem upgrading so long as we can make sure that the newest mlpack is still compilable on recent Ubuntu LTS versions and supported RHEL versions (RHEL7 is current but from mid-2015, probably most departments have been able to transition from RHEL6; RHEL8 is just becoming stable now but we would want to wait for a couple years before having that be required to compile mlpack).  However I haven't taken a look into where exactly compiler support is on each of those distros (and how safe any of these features are to actually try to use on MSVS or other compilers...).

lozhnikov commented:
@rcurtin Thanks for the response! Yes, `std::optional` is a C++17 feature. We can't use it until Ubuntu LTS and RHEL support modern compilers. However, we can use `boost::optional`. If I am not mistaken Boost has the `optional` class since Boost 1.57.

lozhnikov commented:
@rcurtin , @zoq I think the PR is ready. Currently it contains only dictionary encoding. The `Tf-Idf` algorithm (#1969) isn't ready yet. I am going to perform some benchmarks as soon as Jeffin finishes `Tf-Idf`. So, it'll take some time.

Do we want to see this PR in the upcoming release? I am not quite sure that it has any sense without `Tf-Idf`.

zoq commented:
Agreed, this makes more sense in combination with ``Tf-Idf``.

zoq commented on src/mlpack/core/data/string_encoding_impl.hpp:
```diff
@@ -0,0 +1,208 @@
+/**
+ * @file string_encoding_impl.hpp
+ * @author Jeffin Sam
```
I think this was a team effort, so maybe it would be fair to list @lozhnikov as author as well?

lozhnikov replied:
I agree.

zoq commented on src/mlpack/core/data/tokenizers/split_by_any_of.hpp:
```diff
@@ -0,0 +1,113 @@
+   *
+   * @param str String where to find the character.
+   */
+  size_t FindFirstDelimiter(boost::string_view str) const
```
I'm wondering if it makes sense to use ``const boost::string_view& str`` here?

lozhnikov replied:
I think the present implementation could be a bit faster since the size of `boost::string_view` is equal to 16 bytes and the copy-constructor is trivial.

zoq commented on src/mlpack/core/data/string_encoding_policies/dictionary_encoding_policy.hpp:
```diff
@@ -0,0 +1,114 @@
+  * @param dictionarySize The size of the dictionary (not used).
+  */
+  template<typename MatType>
+  static void InitMatrix(MatType& output,
```
Would probably use ``const size_t`` just to be more clear that this parameter isn't going to be changed.

lozhnikov replied:
Fixed everywhere.

lozhnikov commented on src/mlpack/core/data/tokenizers/split_by_any_of.hpp:
```diff
@@ -0,0 +1,114 @@
+   *
+   * @param str String where to find the character.
+   */
+  size_t FindFirstDelimiter(const boost::string_view str) const
```
@zoq Thanks for the review. I pass the argument without reference since I think the present implementation could be a bit faster since the size of `boost::string_view` is equal to 16 bytes and the copy-constructor is trivial.

zoq replied:
Thanks for the clarification, makes sense.

rcurtin replied:
I made a comment on the exact same thing but then I found your response here.  Sounds good to me.  Even if it is not actually faster, with only 16 bytes, it definitely will not be a bottleneck. :+1:

rcurtin commented on src/mlpack/tests/CMakeLists.txt:
```diff
@@ -110,7 +111,6 @@ add_executable(mlpack_test
   timer_test.cpp
   tree_test.cpp
   tree_traits_test.cpp
-  scaling_test.cpp
```
Any reason to remove this test?  Or did I miss something?

lozhnikov replied:
Nice catch! Looks like an incorrect merge.

lozhnikov replied:
Fixed.

rcurtin replied:
Hmm, maybe I overlooked something but it seems like this test is still removed in the diff.

rcurtin commented on src/mlpack/tests/string_encoding_test.cpp:
```diff
@@ -0,0 +1,441 @@
+    "implement a wide array of machine learning methods and functions "
+    "as a \"swiss army knife\" for machine learning researchers.",
+    "In addition to its powerful C++ interface, mlpack also provides "
+    "command-line programs and Python bindings."
```
Might be worth making things a bit harder by adding some Unicode characters, just to make sure there aren't any issues with those.  (I doubt it, I think all the code is fine as is, but never hurts to check. :))

lozhnikov replied:
You're right, I'll work on it.

lozhnikov commented on src/mlpack/core/data/string_encoding_dictionary.hpp:
```diff
@@ -0,0 +1,347 @@
+namespace mlpack {
+namespace data {
+
+/*
```
Yes, you're right. I didn't see it.

lozhnikov replied:
Fixed.

lozhnikov commented on src/mlpack/core/data/string_encoding_dictionary.hpp:
```diff
@@ -0,0 +1,347 @@
+#include <mlpack/core/boost_backport/boost_backport_string_view.hpp>
+#include <unordered_map>
+#include <deque>
+#include <array>
```
`std::array` and `std::deque` aren't used. However, `std::unordered_map` is used a few times. Probably it's better to leave as is in order to minimize the prepocessing time.

lozhnikov commented on src/mlpack/core/data/string_encoding_dictionary.hpp:
```diff
@@ -0,0 +1,347 @@
+  template<typename Archive>
+  void serialize(Archive& ar, const unsigned int /* version */)
+  {
+    ar & BOOST_SERIALIZATION_NVP(mapping);
```
Fixed.

lozhnikov commented on src/mlpack/tests/string_encoding_test.cpp:
```diff
@@ -0,0 +1,441 @@
+
+  const MapType& expectedMapping = expected.Mapping();
+  const MapType& mapping = obtained.Mapping();
+
```
Fixed.

lozhnikov commented on src/mlpack/core/data/string_encoding_impl.hpp:
```diff
@@ -0,0 +1,209 @@
+    EncodingPolicyType encodingPolicy) :
+    encodingPolicy(std::move(encodingPolicy))
+{
+}
```
Fixed everywhere.

jeffin143 commented:
s:keep-open

rcurtin commented:
@lozhnikov it seems like this one got overlooked---I didn't see your responses to the PR comments until now.  Thanks for handling that.  Should I go ahead and merge this or did you have anything else that you wanted to handle before then?

lozhnikov commented:
@rcurtin I fixed a few issues you pointed out. However I think I had forgotten about a feature called `n-gram` (https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.TfidfVectorizer.html). I think it could be quite useful. I'd like to make sure our API allows to implement it. Then I'll merge the PR. Unfortunately I hadn't time to dig into it.

rcurtin commented:
Ok, sounds good.  I think it's okay if we merge this now and then add the n-gram support in a separate PR, but however you want to do it sounds good to me. :+1:

rcurtin commented:
Hey there @lozhnikov, just wanted to check in on this one and see what you think.  I'm still excited about it, but I think we've all been busy over the past months.  Do you think that we should merge as-is, and then open an issue for the n-gram support?  Let me know how I can help out. :+1:

lozhnikov commented:
# N-gram bag of words encoders comparison

I implemented a number of tests. **Sources:** [ngram-tests.zip](https://github.com/mlpack/mlpack/files/4246042/ngram-tests.zip)
As usual I used *Les Miserables* (in plain text) as the input data.

I implemented the following algorithms:
1. **Naive**. It means that each n-gram is a string (ordinary `std::string`) which contains space separated tokens. The algorithm requires a special tokenizer. Other classes weren't changed.
2. **Deque**. The algorithm looks like the previous one. The tokenizer returns a `std::string`, but it uses a `std::deque` in order to reduce the number of token extractions. It works like a conveyor.
3. **Vector**. Each n-gram is a `std::vector<boost::string_view>`. It requires a special dictionary. The dictionary key is a `std::vector<boost::string_view>`. Unfortunately since each `string_view` refers to a string we have to store `std::unordered_set<boost::string_view>` and `std::deque<std::string>` in order to avoid `string_view` invalidation. The set guarantees that we store each token only once.
4. **Vector V2**. The algorithm is similar to the previous one. Again, each n-gram is a `std::vector<boost::string_view>`. The dictionary differs a bit. The dictionary key is a `std::vector<size_t>` where each element of the vector is the ID of the corresponding token. Unfortunately we have to store a map in order to get token IDs (`std::unordered_map<boost::string_view, size_t>`).
5. **Vector V3**. Again, each n-gram is a `std::vector<boost::string_view>`. The dictionary key is a `std::vector<boost::string_view>`. In order to avoid `string_view` invalidation we add each token into a `std::deque<std::string>` and into the dictionary as a vector of length 1.
6. **Static**. The algorithm is the same as **Vector**. The dictionary key is a `std::array<boost::string_view, N>`.
7. **StrView**. Each n-gram is an ordinary `boost::string_view`. The algorithm doesn't ignore the delimiters, therefore two identical n-grams divided by different delimiters are mapped to different numbers.
8. **StrView V2**. Again each n-gram is an ordinary `boost::string_view`, but the hash function and the equality operator ignore the delimiters.

### n-gram range: (1, 1)
![g++_encode_benchmark-(1,1)](https://user-images.githubusercontent.com/10090844/75178532-51832600-5749-11ea-8a68-78c98917d686.png)

### n-gram range: (1, 3)
![g++_encode_benchmark-(1,3)](https://user-images.githubusercontent.com/10090844/75178537-5516ad00-5749-11ea-8d7d-731abf8159fb.png)

### n-gram range: (1, 5)
![g++_encode_benchmark-(1,5)](https://user-images.githubusercontent.com/10090844/75178544-5a73f780-5749-11ea-9542-5252a4f07c1d.png)

### n-gram range: (1, 7)
![g++_encode_benchmark-(1,7)](https://user-images.githubusercontent.com/10090844/75178551-5d6ee800-5749-11ea-9df1-39ada4cb61e4.png)

### n-gram range: (3, 3)
![g++_encode_benchmark-(3,3)](https://user-images.githubusercontent.com/10090844/75178556-6069d880-5749-11ea-974e-0bc65809c1ba.png)

### n-gram range: (3, 5)
![g++_encode_benchmark-(3,5)](https://user-images.githubusercontent.com/10090844/75178572-65c72300-5749-11ea-9e3c-198c95b298fc.png)

### n-gram range: (3, 7)
![g++_encode_benchmark-(3,7)](https://user-images.githubusercontent.com/10090844/75178579-695aaa00-5749-11ea-89e8-5b69e8db7d10.png)

### n-gram range: (5, 5)
![g++_encode_benchmark-(5,5)](https://user-images.githubusercontent.com/10090844/75178586-6bbd0400-5749-11ea-9679-708c1fc98f8e.png)

### n-gram range: (5, 7)
![g++_encode_benchmark-(5,7)](https://user-images.githubusercontent.com/10090844/75178593-6eb7f480-5749-11ea-9828-2e290630c2b7.png)

### n-gram range: (7, 7)
![g++_encode_benchmark-(7,7)](https://user-images.githubusercontent.com/10090844/75178600-711a4e80-5749-11ea-8ddf-9ee79d05bb57.png)

## Conclusion

1. I didn't expect, but an ordinary `std::string` works well with n-grams. I thought `std::vector<boost::string_view>` (or `std::array<boost::string_view, N>`) should work better but it's not the case.
2. Unfortunately, `std::deque` doesn't make any significant difference.
3. The last algorithm (**StrView V2**) sometimes outperforms the **Naive** approach. But the difference isn't significant.

Anyway, it is possible to implement the n-grams support in terms of the API.

lozhnikov commented:
@rcurtin Sorry for the slow response. I promised to sort it out in October-November, but things didn't go as planned.

I've done a small investigation. The n-gram support doesn't break the API. Perhaps it's a good idea to make `MapType` a template parameter in `StringEncodingDictionary`, but we can do it later since it's an internal part of the API. And I'm not 100% sure it's really needed.

I think we should move the n-gram support into another PR otherwise we'll never merge this one:). I'm a bit surprised that an ordinary `std::string` works fine with n-grams. I thought `std::array<boost::string_view>` should outperform ordinary strings like `string_view`s do. But it's not the case. I'd like to dig into it a bit more.

jeffin143 commented:
You are a hero @lozhnikov :)

lozhnikov commented:
@jeffin143 Thanks:)

@rcurtin I think the PR is ready now. I added a couple of unicode symbols (several emojis) to the test. Looks like everything works fine. However, we don't have a proper unicode support without wide characters. It's easy to add the wide strings support, but I'm not sure it's needed since multibyte encodings like UTF-8 are more popular than fixed-size Unicode. Anyway, this PR is ready. So, let's wait for the results of the automated checks.

rcurtin commented on src/mlpack/tests/string_encoding_test.cpp:
```diff
@@ -0,0 +1,524 @@
+  }
+
+  arma::mat expected = {
+    { 1,  2, 0, 0, 0 },
```
```suggestion
    { 1, 2, 0, 0, 0 },
```
Super pedantic comment. :)

lozhnikov replied:
I've already added `scaling_test.hpp`. See line 75. Now the lines are in alphabetic order.

rcurtin replied:
Weird, I think that I was looking at the wrong diff---it looks like I wasn't seeing all changes.  It seems fine, sorry about that. :+1:

lozhnikov commented:
> Really cool results with the n-gram experiments; I was surprised that the naive approach performed so well! Anyway, I do agree, perhaps better to put that in a different PR. :)

Looks like other approaches overcomplicate the dictionary. I think they increase the number of lookup operations. I found out recently that `std::string` stores short strings (up to 15 or 16 bytes if I am not mistaken) on the stack. It's interesting. I didn't know.

> Everything looks great to me, just a couple tiny comments I found when passing through the diff. Should we merge this once it passes the CI tests?

I removed the space. Yes, I think we can merge it.

rcurtin commented:
Perfect, sounds good.  The AppVeyor build might take a little while since there are so many PRs building these days, but I imagine it will pass, then we can merge it. :+1:
